packages/tangle/scheduler.go

```diff
@@ -279,29 +280,33 @@ func (s *Scheduler) schedule() *Message {
 	s.mu.Lock()
 	defer s.mu.Unlock()
 
+	// cache the access mana
+	manaCache := s.tangle.Options.SchedulerParams.AccessManaMapRetrieverFunc()
+
+	s.updateActiveNodesList(manaCache)
+
 	start := s.buffer.Current()
 	// no messages submitted
 	if start == nil {
 		return nil
 	}
 
-	// cache the access mana retrieval
-	manas := make(map[identity.ID]float64, s.buffer.NumActiveNodes())
 	getCachedMana := func(id identity.ID) float64 {
-		if mana, ok := manas[id]; ok {
+		if mana, ok := manaCache[id]; ok {
 			return mana
 		}
-		mana := math.Max(s.tangle.Options.SchedulerParams.AccessManaRetrieveFunc(id), MinMana)
-		manas[id] = mana
+		// AccessManaRetrieveFunc always returns at least MinMana
+		mana := s.tangle.Options.SchedulerParams.AccessManaRetrieveFunc(id)
```

Comment:
> I don't know should we rather panic in this case? Then hiding this error?

Currently in `AccessBaseManaVector.getMana` error is never returned. It could be returned by `GetAccessMana` in mana plugin, if querying mana plugin is not allowed, however currently query is always allowed (hardcoded `true`). So if we hide this error then nothing gets scheduled, or all nodes are assumed to have MinMana (after I moved the hardcoded MinMana to the scheduler) and things are scheduled.
I think we can leave it as it is as the error should never be returned anyway.

Reply:
The check for zero mana is added.

Reply:
> Couldn't there be a zero-mana node that still has a pending message? 

This is an interesting case (after we remove the hard-coded MinMana assignment). Originally I was thinking that `updateActiveNodesList` should remove pending messages of nodes that have less than MinMana, because they might never accumulate enough deficit again. WDYT?

Reply:
I've add this, please check it out